File: pkg/bench_nav/src/bench_nav/stepped/bug/mem_astar.py
"""Memory+A* — stepped.

Structurally identical to Memory+BFS but expands cells by f = g + h
(Chebyshev), so fewer expansions reach the same depth; same expansion
budget buys a longer effective horizon.
"""

from __future__ import annotations

import heapq
from enum import Enum
from typing import override

from bench_nav.common import INF
from bench_nav.precomputation import COST
from bench_nav.stepped.bug._common import (
    WallFollowState,
    WallStepOutcome,
    bresenham,
    dir_to_goal,
    dist_sq,
    has_los,
    make_passable_closures,
    neighbour,
    sensed_cells,
    wall_follow_step,
)
from bench_nav.types import PrecompCtx, Stepped

_EXPAND_BUDGET = 500
_DIST_INF = 65535
_DIRS8: tuple[tuple[int, int], ...] = (
    (1, 0),
    (-1, 0),
    (0, 1),
    (0, -1),
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
)


class _Mode(Enum):
    MOTION = 0
    CIRCUMNAV = 1
    RETURN_TO_LEAVE = 2


class MemAstar(Stepped):
    REQUIRES = frozenset({COST})

    @override
    def __init__(self, ctx: PrecompCtx) -> None:
        self.w = ctx.w
        self.h = ctx.h
        self.cost = ctx[COST]
        self.passable, self.on_map = make_passable_closures(
            ctx.w, ctx.h, self.cost, INF
        )
        self._active_goal: int | None = None
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(
            pos=(0, 0), current_obstacle=(0, 0), obstacle_on_right=True
        )
        self._best_leave: tuple[int, int] = (0, 0)
        self._best_leave_dist_sq = 0
        self._global_min_dist_sq = 0
        self._follow_visited: set[tuple[tuple[int, int], tuple[int, int], bool]] = set()
        self._los_queue: list[tuple[int, int]] = []
        n = ctx.w * ctx.h
        self._pnb: list[list[int]] = [[] for _ in range(n)]
        self._discovered = bytearray(n)
        self._g_arr: list[int] = [_DIST_INF] * n

    def _reset(self, p: tuple[int, int], g: tuple[int, int]) -> None:
        w, h = self.w, self.h
        n = w * h
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(pos=p, current_obstacle=p, obstacle_on_right=True)
        self._best_leave = p
        self._best_leave_dist_sq = dist_sq(p, g)
        self._global_min_dist_sq = dist_sq(p, g)
        self._follow_visited.clear()
        self._los_queue.clear()
        pnb = self._pnb
        for i in range(n):
            pnb[i] = []
        for y in range(h):
            for x in range(w):
                i = y * w + x
                row = pnb[i]
                for dx, dy in _DIRS8:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < w and 0 <= ny < h:
                        row.append(ny * w + nx)
        self._discovered = bytearray(n)

    def _sense(self, cur_pos: tuple[int, int]) -> None:
        w, h = self.w, self.h
        cost = self.cost
        discovered = self._discovered
        pnb = self._pnb
        for c in sensed_cells(cur_pos):
            x, y = c
            if not (0 <= x < w and 0 <= y < h):
                continue
            i = y * w + x
            if discovered[i]:
                continue
            discovered[i] = 1
            if cost[i] >= INF:
                for dx, dy in _DIRS8:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < w and 0 <= ny < h:
                        ni = ny * w + nx
                        if i in pnb[ni]:
                            pnb[ni].remove(i)
                pnb[i].clear()

    def _try_astar_step(
        self, cur_pos: tuple[int, int], goal: tuple[int, int], threshold: int
    ) -> tuple[int, int] | None:
        w = self.w
        n = w * self.h
        pnb = self._pnb
        g_arr = self._g_arr
        gx, gy = goal[0], goal[1]

        def heur(i: int) -> int:
            nx, ny = i % w, i // w
            return max(abs(nx - gx), abs(ny - gy))

        pos_idx = cur_pos[1] * w + cur_pos[0]
        for i in range(n):
            g_arr[i] = _DIST_INF
        g_arr[pos_idx] = 0
        heap: list[tuple[int, int]] = [(heur(pos_idx), pos_idx)]
        expansions = 0
        best_idx: int | None = None
        best_d = threshold
        reached_goal = False
        goal_idx = goal[1] * w + goal[0]

        while heap:
            f, node = heapq.heappop(heap)
            node_g = g_arr[node]
            if node_g == _DIST_INF:
                continue
            if f > node_g + heur(node):
                continue
            expansions += 1
            if expansions > _EXPAND_BUDGET:
                break
            if node == goal_idx:
                reached_goal = True
                best_idx = node
                break
            nx, ny = node % w, node // w
            dx = nx - gx
            dy = ny - gy
            d = dx * dx + dy * dy
            if d < best_d:
                best_d = d
                best_idx = node
            g_new = node_g + 1
            for ni in pnb[node]:
                if g_new < g_arr[ni]:
                    g_arr[ni] = g_new
                    heapq.heappush(heap, (g_new + heur(ni), ni))

        if best_idx is None:
            return None
        cur = best_idx
        while g_arr[cur] > 1:
            cur_g = g_arr[cur]
            next_cell: int | None = None
            for ni in pnb[cur]:
                if g_arr[ni] == cur_g - 1:
                    next_cell = ni
                    break
            if next_cell is None:
                return None
            cur = next_cell
        if g_arr[cur] == _DIST_INF or cur == pos_idx:
            return None
        first_step = (cur % w, cur // w)
        if not reached_goal and dist_sq(first_step, goal) >= threshold:
            return None
        return first_step

    @override
    def step(self, pos: int, goal: int) -> int | None:
        w = self.w
        p = (pos % w, pos // w)
        g = (goal % w, goal // w)
        if goal != self._active_goal:
            self._active_goal = goal
            self._reset(p, g)
        if not self.passable(*p) or not self.passable(*g):
            return None
        if p == g:
            return pos

        if self._los_queue:
            nxt = self._los_queue.pop(0)
            return nxt[1] * w + nxt[0]
        self._sense(p)
        if has_los(p, g, self.passable):
            queue = bresenham(p, g)[1:]
            if queue:
                self._los_queue = queue[1:]
                first = queue[0]
                return first[1] * w + first[0]

        threshold = (
            dist_sq(p, g)
            if self._mode is _Mode.MOTION
            else min(self._global_min_dist_sq, self._best_leave_dist_sq)
        )
        astar_step = self._try_astar_step(p, g, threshold)
        if astar_step is not None:
            d2 = dist_sq(astar_step, g)
            self._global_min_dist_sq = min(self._global_min_dist_sq, d2)
            if self._mode is not _Mode.MOTION:
                self._mode = _Mode.MOTION
                self._follow_visited.clear()
            return astar_step[1] * w + astar_step[0]

        if self._mode is _Mode.MOTION:
            d = dir_to_goal(p, g)
            np = neighbour(p, d)
            if self.passable(*np):
                d2 = dist_sq(np, g)
                self._global_min_dist_sq = min(self._global_min_dist_sq, d2)
                return np[1] * w + np[0]
            self._mode = _Mode.CIRCUMNAV
            self._wf = WallFollowState(
                pos=p,
                current_obstacle=neighbour(p, d),
                obstacle_on_right=True,
            )
            self._best_leave = p
            self._best_leave_dist_sq = dist_sq(p, g)
            self._follow_visited.clear()
            self._follow_visited.add(
                (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
            )
            return pos
        if self._mode is _Mode.CIRCUMNAV:
            self._wf.pos = p
            outcome = wall_follow_step(self._wf, self.passable, self.on_map)
            if outcome is WallStepOutcome.SURROUNDED:
                return None
            np = self._wf.pos
            d2 = dist_sq(np, g)
            if d2 < self._best_leave_dist_sq:
                self._best_leave = np
                self._best_leave_dist_sq = d2
            state = (
                self._wf.pos,
                self._wf.current_obstacle,
                self._wf.obstacle_on_right,
            )
            if state in self._follow_visited:
                if self._best_leave_dist_sq >= self._global_min_dist_sq:
                    return None
                self._global_min_dist_sq = self._best_leave_dist_sq
                self._mode = _Mode.RETURN_TO_LEAVE
                self._follow_visited.clear()
                self._follow_visited.add(state)
            else:
                self._follow_visited.add(state)
            return np[1] * w + np[0]
        # RETURN_TO_LEAVE
        if p == self._best_leave:
            self._mode = _Mode.MOTION
            return pos
        self._wf.pos = p
        outcome = wall_follow_step(self._wf, self.passable, self.on_map)
        if outcome is WallStepOutcome.SURROUNDED:
            return None
        np = self._wf.pos
        state = (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
        if state in self._follow_visited:
            return None
        self._follow_visited.add(state)
        return np[1] * w + np[0]

File: crates/titan-bugnav/src/algorithms/mem_astar.rs
//! Memory+A* — identical structure to Memory+BFS but expands cells in order
//! of f = g + h (Chebyshev heuristic). Goal-directed, so fewer expansions
//! reach the same depth; same BFS budget buys a longer effective horizon.

use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashSet};

use crate::algorithms::bug_common::{
    VISION_R_SQ, WallFollowState, WallStepOutcome, bresenham, dir_to_goal, dist_sq, has_los,
    neighbour, sensed_cells, wall_follow_step,
};
use crate::grid::Grid;
use crate::pathfinder::{Pathfinder, Snapshot, StepStatus};

const EXPAND_BUDGET: u32 = 500;
const DIST_INF: u16 = u16::MAX;

const DIRS8: [(i32, i32); 8] = [
    (1, 0),
    (-1, 0),
    (0, 1),
    (0, -1),
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Mode {
    MotionToGoal,
    Circumnavigate,
    ReturnToLeave,
}

pub struct MemAstar {
    grid_w: i32,
    grid_h: i32,
    walls: Vec<bool>,
    goal: (i32, i32),
    pos: (i32, i32),
    discovered: Vec<bool>,
    pnb: Vec<Vec<u16>>,
    g: Vec<u16>,
    g_reset: Vec<u16>,
    heap: BinaryHeap<Reverse<(u32, u16)>>,
    mode: Mode,
    wf: WallFollowState,
    hit_point: (i32, i32),
    best_leave: (i32, i32),
    best_leave_dist_sq: i32,
    global_min_dist_sq: i32,
    follow_visited: HashSet<((i32, i32), (i32, i32), bool)>,
    steps: u32,
    snap: Snapshot,
    status: StepStatus,
}

#[must_use]
pub fn build(grid: &Grid, start: (i32, i32), goal: (i32, i32)) -> Box<dyn Pathfinder> {
    let w = grid.w;
    let h = grid.h;
    let n = (w * h) as usize;

    let mut pnb: Vec<Vec<u16>> = vec![Vec::with_capacity(8); n];
    for y in 0..h {
        for x in 0..w {
            let i = (y * w + x) as usize;
            for (dx, dy) in DIRS8 {
                let nx = x + dx;
                let ny = y + dy;
                if nx >= 0 && nx < w && ny >= 0 && ny < h {
                    let ni = (ny * w + nx) as u16;
                    pnb[i].push(ni);
                }
            }
        }
    }

    let g_reset = vec![DIST_INF; n];
    let g = g_reset.clone();

    let mut snap = Snapshot {
        current: start,
        path: vec![start],
        ..Snapshot::default()
    };
    snap.visited.insert(start);

    Box::new(MemAstar {
        grid_w: w,
        grid_h: h,
        walls: grid.walls.clone(),
        goal,
        pos: start,
        discovered: vec![false; n],
        pnb,
        g,
        g_reset,
        heap: BinaryHeap::with_capacity(4096),
        mode: Mode::MotionToGoal,
        wf: WallFollowState {
            pos: start,
            current_obstacle: start,
            obstacle_on_right: true,
        },
        hit_point: start,
        best_leave: start,
        best_leave_dist_sq: dist_sq(start, goal),
        global_min_dist_sq: dist_sq(start, goal),
        follow_visited: HashSet::new(),
        steps: 0,
        snap,
        status: StepStatus::Running,
    })
}

impl MemAstar {
    #[inline]
    const fn idx(&self, x: i32, y: i32) -> Option<usize> {
        if x < 0 || y < 0 || x >= self.grid_w || y >= self.grid_h {
            None
        } else {
            Some((y * self.grid_w + x) as usize)
        }
    }

    #[inline]
    const fn cell_of(&self, i: usize) -> (i32, i32) {
        let w = self.grid_w;
        ((i as i32) % w, (i as i32) / w)
    }

    #[inline]
    fn h(&self, idx: usize) -> u32 {
        let (nx, ny) = self.cell_of(idx);
        let dx = (nx - self.goal.0).unsigned_abs();
        let dy = (ny - self.goal.1).unsigned_abs();
        dx.max(dy)
    }

    fn sense(&mut self) {
        for c in sensed_cells(self.pos) {
            let Some(i) = self.idx(c.0, c.1) else {
                continue;
            };
            if self.discovered[i] {
                continue;
            }
            self.discovered[i] = true;
            if self.walls[i] {
                let (x, y) = self.cell_of(i);
                let i_u16 = i as u16;
                for (dx, dy) in DIRS8 {
                    let nx = x + dx;
                    let ny = y + dy;
                    if nx >= 0 && nx < self.grid_w && ny >= 0 && ny < self.grid_h {
                        let ni = (ny * self.grid_w + nx) as usize;
                        self.pnb[ni].retain(|&c| c != i_u16);
                    }
                }
                self.pnb[i].clear();
            }
        }
    }

    fn passable(&self, x: i32, y: i32) -> bool {
        let Some(i) = self.idx(x, y) else {
            return false;
        };
        !self.walls[i]
    }

    fn move_to(&mut self, new_pos: (i32, i32)) {
        self.pos = new_pos;
        self.snap.current = new_pos;
        self.snap.visited.insert(new_pos);
        self.snap.path.push(new_pos);
    }

    /// Bounded A* over `pnb` starting at `self.pos`. Returns the first-step
    /// cell along the path to the cell that minimises `dist_sq(c, goal)`
    /// subject to `< threshold`. Unconditionally commits if goal reached.
    fn try_astar_step(&mut self, threshold: i32) -> bool {
        let pos_idx = self
            .idx(self.pos.0, self.pos.1)
            .expect("pos must be in-bounds");

        self.g.copy_from_slice(&self.g_reset);
        self.g[pos_idx] = 0;

        self.heap.clear();
        self.heap.push(Reverse((self.h(pos_idx), pos_idx as u16)));

        let mut expansions: u32 = 0;
        let mut best_idx: Option<usize> = None;
        let mut best_d: i32 = threshold;
        let mut reached_goal = false;
        let goal_idx = self.idx(self.goal.0, self.goal.1);

        while let Some(Reverse((f, node_u16))) = self.heap.pop() {
            let node = node_u16 as usize;
            let node_g = self.g[node];
            if node_g == DIST_INF {
                continue;
            }
            // Stale entry (node was relaxed after this was pushed).
            if f as u16 > node_g.saturating_add(self.h(node) as u16) {
                continue;
            }

            expansions += 1;
            if expansions > EXPAND_BUDGET {
                break;
            }

            if Some(node) == goal_idx {
                reached_goal = true;
                best_idx = Some(node);
                break;
            }
            let (nx, ny) = self.cell_of(node);
            let dx = nx - self.goal.0;
            let dy = ny - self.goal.1;
            let d = dx * dx + dy * dy;
            if d < best_d {
                best_d = d;
                best_idx = Some(node);
            }

            let g_new = node_g + 1;
            let pnb_len = self.pnb[node].len();
            for k in 0..pnb_len {
                let ni_u16 = self.pnb[node][k];
                let ni = ni_u16 as usize;
                if g_new < self.g[ni] {
                    self.g[ni] = g_new;
                    let f_ni = u32::from(g_new) + self.h(ni);
                    self.heap.push(Reverse((f_ni, ni_u16)));
                }
            }
        }

        let Some(target) = best_idx else {
            return false;
        };

        // Gradient-descent on g back to a cell adjacent to pos.
        let mut cur = target;
        while self.g[cur] > 1 {
            let cur_g = self.g[cur];
            let mut next_cell: Option<usize> = None;
            for &ni_u16 in &self.pnb[cur] {
                let ni = ni_u16 as usize;
                if self.g[ni] == cur_g - 1 {
                    next_cell = Some(ni);
                    break;
                }
            }
            let Some(n) = next_cell else {
                return false;
            };
            cur = n;
        }
        if self.g[cur] == DIST_INF {
            return false;
        }
        if cur == pos_idx {
            return false;
        }
        let first_step = self.cell_of(cur);

        if !reached_goal {
            let fs_d = dist_sq(first_step, self.goal);
            if fs_d >= threshold {
                return false;
            }
        }

        self.move_to(first_step);
        let d2 = dist_sq(self.pos, self.goal);
        if d2 < self.global_min_dist_sq {
            self.global_min_dist_sq = d2;
        }
        true
    }

    fn step_wall_follow(&mut self) -> WallStepOutcome {
        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable =
            |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize];
        let on_map = |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h;
        let outcome = wall_follow_step(&mut self.wf, passable, on_map);
        if outcome == WallStepOutcome::Moved {
            self.move_to(self.wf.pos);
        }
        outcome
    }

    fn enter_circumnavigate(&mut self, blocked_dir: usize) {
        self.mode = Mode::Circumnavigate;
        self.wf = WallFollowState {
            pos: self.pos,
            current_obstacle: neighbour(self.pos, blocked_dir),
            obstacle_on_right: true,
        };
        self.hit_point = self.pos;
        self.best_leave = self.pos;
        self.best_leave_dist_sq = dist_sq(self.pos, self.goal);
        self.follow_visited.clear();
        self.follow_visited.insert((
            self.wf.pos,
            self.wf.current_obstacle,
            self.wf.obstacle_on_right,
        ));
    }
}

impl Pathfinder for MemAstar {
    fn step(&mut self) -> StepStatus {
        if self.status != StepStatus::Running {
            return self.status;
        }
        if self.pos == self.goal {
            self.status = StepStatus::Arrived;
            return self.status;
        }
        self.steps += 1;

        self.sense();

        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable_closure = move |x: i32, y: i32| {
            x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize]
        };
        if has_los(self.pos, self.goal, passable_closure) {
            let line = bresenham(self.pos, self.goal);
            for &p in line.iter().skip(1) {
                self.move_to(p);
            }
            self.status = StepStatus::Arrived;
            return self.status;
        }

        let threshold = match self.mode {
            Mode::MotionToGoal => dist_sq(self.pos, self.goal),
            Mode::Circumnavigate | Mode::ReturnToLeave => {
                self.global_min_dist_sq.min(self.best_leave_dist_sq)
            }
        };
        if self.try_astar_step(threshold) {
            if self.mode != Mode::MotionToGoal {
                self.mode = Mode::MotionToGoal;
                self.follow_visited.clear();
            }
            return self.status;
        }

        match self.mode {
            Mode::MotionToGoal => {
                let d = dir_to_goal(self.pos, self.goal);
                let np = neighbour(self.pos, d);
                if self.passable(np.0, np.1) {
                    self.move_to(np);
                    let d2 = dist_sq(self.pos, self.goal);
                    if d2 < self.global_min_dist_sq {
                        self.global_min_dist_sq = d2;
                    }
                } else {
                    self.enter_circumnavigate(d);
                }
            }
            Mode::Circumnavigate => {
                match self.step_wall_follow() {
                    WallStepOutcome::Moved => {}
                    WallStepOutcome::Surrounded => {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                }
                let d2 = dist_sq(self.pos, self.goal);
                if d2 < self.best_leave_dist_sq {
                    self.best_leave = self.pos;
                    self.best_leave_dist_sq = d2;
                }
                let state = (
                    self.wf.pos,
                    self.wf.current_obstacle,
                    self.wf.obstacle_on_right,
                );
                if !self.follow_visited.insert(state) {
                    if self.best_leave_dist_sq >= self.global_min_dist_sq {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                    self.global_min_dist_sq = self.best_leave_dist_sq;
                    self.mode = Mode::ReturnToLeave;
                    self.follow_visited.clear();
                    self.follow_visited.insert((
                        self.wf.pos,
                        self.wf.current_obstacle,
                        self.wf.obstacle_on_right,
                    ));
                }
            }
            Mode::ReturnToLeave => {
                if self.pos == self.best_leave {
                    self.mode = Mode::MotionToGoal;
                    return self.status;
                }
                match self.step_wall_follow() {
                    WallStepOutcome::Moved => {}
                    WallStepOutcome::Surrounded => {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                }
                let state = (
                    self.wf.pos,
                    self.wf.current_obstacle,
                    self.wf.obstacle_on_right,
                );
                if !self.follow_visited.insert(state) {
                    self.status = StepStatus::Unreachable;
                }
            }
        }
        self.status
    }

    fn snapshot(&self) -> &Snapshot {
        &self.snap
    }

    fn summary(&self) -> String {
        let discovered_count = self.discovered.iter().filter(|&&d| d).count();
        let total = self.discovered.len();
        format!(
            "pos: ({}, {})\ngoal: ({}, {})\nmode: {:?}\ndiscovered: {}/{} ({:.1}%)\nA* budget: {}\nsensor r²≤{}\nsteps: {}\nstatus: {:?}",
            self.pos.0,
            self.pos.1,
            self.goal.0,
            self.goal.1,
            self.mode,
            discovered_count,
            total,
            100.0 * discovered_count as f64 / total as f64,
            EXPAND_BUDGET,
            VISION_R_SQ,
            self.steps,
            self.status,
        )
    }

    fn name(&self) -> &'static str {
        "Memory+A*"
    }
}
